File: 2023/03/part-2.lua
input_path = "input"
sum = 0

lines = {}
for line in io.lines(input_path) do
  table.insert(lines,line)
end

function add_numbers_from_line(numbers,str,index)
  if str == nil then
    return
  end

  for i, numeric, j in str:gmatch("()(%d+)()") do
    if i <= index + 1 and j-1 >= index - 1 then
      table.insert(numbers,tonumber(numeric))
    end
  end
end

for index,line in ipairs(lines) do
  for i,_ in line:gmatch("()(%*)") do
    numbers = {}
    add_numbers_from_line(numbers, lines[index - 1], i)
    add_numbers_from_line(numbers, line, i)
    add_numbers_from_line(numbers, lines[index + 1], i)

    if #numbers == 2 then
      sum = sum + (numbers[1] * numbers[2])
    end
  end
end

-- Expected for sample: 4361
print("Result: " .. sum)

File: 2024/07/part_2.rb
require_relative('../../lib/support')

solve example: 11387 do |input|
  operations = [:*, :+, :concat]

  input.lines.flat_map do |line|
    total, number_string = line.split(': ')
    total = total.to_i
    numbers = number_string.split(' ').map(&:to_i)

    permutations = operations.repeated_permutation(numbers.size - 1)
    total if permutations.any? do |permutation|
      result = numbers[0]
      (1...numbers.size).each do |index|
        operation = permutation[index - 1]
        number = numbers[index]

        case operation
        when :+
          result += number
        when :*
          result *= number
        when :concat
          result = (result.to_s + number.to_s).to_i
        end
      end

      result == total
    end
  end.compact.sum
end
